modules/component-store/src/component-store.ts

```diff
@@ -0,0 +1,114 @@
+  private readonly destroySubject$ = new ReplaySubject<void>(1);
+  // Exposed to any extending Store to be used for the teardowns.
+  readonly destroy$ = this.destroySubject$.asObservable();
+
+  constructor(defaultState?: T) {
+    // State can be initialized either through constructor, or initState or
+    // setState.
+    if (defaultState) {
+      this.initState(defaultState);
+    }
+  }
+
+  /** Completes all relevant Observable streams. */
+  ngOnDestroy() {
+    this.stateSubject$.complete();
+    this.destroySubject$.next();
+  }
+
+  /**
+   * Creates an updater.
+   *
+   * Throws an error if updater is called with synchronous values (either
+   * imperative value or Observable that is synchronous) before ComponentStore
+   * is initialized. If called with async Observable before initialization then
+   * state will not be updated and subscription would be closed.
+   *
+   * @param updaterFn A static updater function that takes 2 parameters (the
+   * current state and an argument object) and returns a new instance of the
+   * state.
+   * @return A function that accepts one argument which is forwarded as the
+   *     second argument to `updaterFn`. Everytime this function is called
+   *     subscribers will be notified of the state change.
+   */
+  updater<V>(
+    updaterFn: (state: T, value: V) => T
+  ): unknown extends V ? () => void : (t: V | Observable<V>) => Subscription {
+    return ((observableOrValue?: V | Observable<V>): Subscription => {
+      let initializationError: Error | undefined;
+      // We can receive either the value or an observable. In case it's a
+      // simple value, we'll wrap it with `of` operator to turn it into
+      // Observable.
+      const observable$ = isObservable(observableOrValue)
+        ? observableOrValue
+        : of(observableOrValue);
+      const subscription = observable$
+        .pipe(
+          concatMap(
+            value =>
+              this.isInitialized
+                ? of(value).pipe(withLatestFrom(this.stateSubject$))
+                : // If state was not initialized, we'll throw an error.
+                  throwError(
+                    Error(`${this.constructor.name} has not been initialized`)
+                  )
+          ),
+          takeUntil(this.destroy$)
+        )
+        .subscribe({
+          next: ([value, currentState]) => {
+            this.stateSubject$.next(updaterFn(currentState, value!));
+          },
+          error: error => {
+            initializationError = error;
+            this.stateSubject$.error(error);
+          },
+        });
+
+      if (initializationError) {
+        throw initializationError;
+      }
+      return subscription;
+    }) as unknown extends V
+      ? () => void
+      : (t: V | Observable<V>) => Subscription;
+  }
+
+  /**
+   * Initializes state. If it was already initialized then it resets the
+   * state.
+   */
+  private initState(state: T): void {
```

Comment:
Nit, I find the name `initState` a bit missleading because it can also be invoked via `setState`.
Should we rename this function to `updateState`/`resetState`?

Reply:
it could be `resetState`, however it's also the only part that can set `this.isInitialized = true`, which might be more important piece of the info.

I'm fine either way.
WDYT?

Reply:
I'm fine with keeping it, the comment was mostly based of the previous commit.

Reply:
I see.
Since it's also private now, we can revisit this later if necessary 🙂 (should we decide to open this method up...)